PowerEditor/src/NppIO.cpp

```diff
@@ -93,7 +94,12 @@ DWORD WINAPI Notepad_plus::monitorFileOnChange(void * params)
 					{
 						if (dwAction == FILE_ACTION_MODIFIED)
 						{
-							::PostMessage(h, NPPM_INTERNAL_RELOADSCROLLTOEND, reinterpret_cast<WPARAM>(buf), 0);
+							if (!bDoneOnce)
+							{
+								::PostMessage(h, NPPM_INTERNAL_RELOADSCROLLTOEND, reinterpret_cast<WPARAM>(buf), 0);
+								bDoneOnce = true;
+								Sleep(250);

```

Comment:
Where come from the magic number "250" ?

Reply:
Consider the preceding line:
`69 	DWORD waitStatus = ::WaitForMultipleObjects(_countof(changeHandles), changeHandles, FALSE, 250);`

WaitForMultipleObjects waits for file modification events, or times out after 250ms, acting as a polling loop.

But, when monitoring log files actively being written, hundreds of events can be returned in one call, and each matching FILE_ACTION_MODIFIED event will PostMessage(h, NPPM_INTERNAL_RELOADSCROLLTOEND... leading to a freeze.
```
	if (!bDoneOnce)
	{
		::PostMessage(h, NPPM_INTERNAL_RELOADSCROLLTOEND, reinterpret_cast<WPARAM>(buf), 0);
		bDoneOnce = true;
		Sleep(250);
	}
```
Above, I am using a boolean flag to only PostMessage for first matching event, and then Sleeping for a similar 250ms as used in the polling loop, to break the flood of reloads.

Reply:
I don't get the complete logic here... Can you explain? Why is both needed? The boolean flag to trigger PostMessage only once and the sleep afterwards?

If the PostMessage is triggered only once, the application does not freeze anymore. Only first event is processed and the all others will be ignored.

Sleep(...) blocks the application from running, which seems not needed (could even lead to non-responsive app, too). IMO Sleep(...) should be avoided.

Reply:
No, it does not block the UI, even if you use a delay like 10 seconds. It simply limits the refresh rate for a monitored document with rapid continuous updates. It is without it that the UI becomes unresponsive. 

Just because everyone has an opinion on Sleep(...) doesn't mean this approach is not correct here.

Reply:
Thank you for clarifiing and confirming, that it will not block the application. That's fine.

But still don't get completely why calling Sleep(...) is needed at all.

As from my understanding, there are multiple FILE_ACTION_MODIFIED which get popped from the available dirChanges (in while loop). And all of them would invoke a PostMessage(... NPPM_INTERNAL_RELOADSCROLLTOEND ...) which is the cause of Notepad++ becoming unresponsive for a while.

But since with your change, PostMessage(... NPPM_INTERNAL_RELOADSCROLLTOEND ...) is invoked only once (the first time a FILE_ACTION_MODIFIED is received), all further FILE_ACTION_MODIFIED will be ignored anyway, even without calling Sleep(...). So the root cause of unresponsiveness seems gone, even without calling Sleep(...).

Limiting the refresh rate seems already been done by WaitForMultipleObjects(...) because it also waits for 250ms.

Not sure if I have formulated my thoughts well, nor whether they are correct. Will try to test a bit.

Reply:
@MarkusBodensee WaitForMultipleObjects rarely waits the full 250ms, unless there are no events queued. Once the PostMessage is sent, editor will get busy with loading, parsing and layout and whatever else with the plugins, all this while more events accumulate. If there is no breathing space between each refresh, it become unresponsive because it is busy all the time. It gets worse as the file grows.

To test under load, run a command like `for /l %x in (1,1,100000) do echo line number %x >> testfile.txt` in a windows command prompt, and then load and monitor the file testfile.txt while it is being continuously updated.

Reply:
Thank you for your explanations.